cmd/opampsupervisor/supervisor/supervisor.go

```diff
@@ -0,0 +1,695 @@
+
+	s.createInstanceID()
+	logger.Debug("Supervisor starting",
+		zap.String("id", s.instanceID.String()), zap.String("type", agentType), zap.String("version", s.agentVersion))
+
+	s.loadAgentEffectiveConfig()
+
+	if err := s.startOpAMP(); err != nil {
+		return nil, fmt.Errorf("cannot start OpAMP client: %w", err)
+	}
+
+	s.commander, err = commander.NewCommander(
+		s.logger,
+		s.config.Agent,
+		"--config", s.effectiveConfigFilePath,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	s.startHealthCheckTicker()
+	go s.runAgentProcess()
+
+	return s, nil
+}
+
+func (s *Supervisor) loadConfig(configFile string) error {
+	if configFile == "" {
+		return errors.New("path to config file cannot be empty")
+	}
+
+	k := koanf.New("::")
+	if err := k.Load(file.Provider(configFile), yaml.Parser()); err != nil {
+		return err
+	}
+
+	if err := k.Unmarshal("", &s.config); err != nil {
+		return fmt.Errorf("cannot parse %v: %w", configFile, err)
+	}
+
+	return nil
+}
+
+func (s *Supervisor) getBootstrapInfo() (err error) {
+	s.agentVersion = "1.0.0"
+
+	return nil
+}
+
+func (s *Supervisor) startOpAMP() error {
+	s.opampClient = client.NewWebSocket(opAMPLogger{logger: s.logger})
+
+	settings := types.StartSettings{
+		OpAMPServerURL: s.config.Server.Endpoint,
+		InstanceUid:    s.instanceID.String(),
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc: func() {
+				s.logger.Debug("Connected to the server.")
+			},
+			OnConnectFailedFunc: func(err error) {
+				s.logger.Error("Failed to connect to the server", zap.Error(err))
+			},
+			OnErrorFunc: func(err *protobufs.ServerErrorResponse) {
+				s.logger.Error("Server returned an error response", zap.String("message", err.ErrorMessage))
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				return s.createEffectiveConfigMsg(), nil
+			},
+			OnMessageFunc: s.onMessage,
+		},
+		// TODO: Make capabilities configurable
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_AcceptsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsEffectiveConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsOwnMetrics |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsHealth,
+	}
+	err := s.opampClient.SetAgentDescription(s.createAgentDescription())
+	if err != nil {
+		return err
+	}
+
+	err = s.opampClient.SetHealth(&protobufs.AgentHealth{Healthy: false})
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debug("Starting OpAMP client...")
+
+	err = s.opampClient.Start(context.Background(), settings)
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debug("OpAMP Client started.")
+
+	return nil
+}
+
+func (s *Supervisor) createInstanceID() {
+	// Generate instance id.
+	entropy := ulid.Monotonic(rand.New(rand.NewSource(0)), 0)
+	s.instanceID = ulid.MustNew(ulid.Timestamp(time.Now()), entropy)
+
+	// TODO: Persist instance ID.
```

Comment:
Please add a link to appropriate section in design doc: https://docs.google.com/document/d/1KtH5atZQUs9Achbce6LiOaJxLbksNJenvgvyKLsJrkc/edit#heading=h.8szgxmjard0e

Reply:
Thanks for the link. I've created an issue and linked to it here so we have something to track the work.